### PR TITLE
resource listing

### PR DESCRIPTION
**Pull request by adithyadinesh0412**

### CODE REVIEW

File: src/database/migrations/20240531070905-alter-resource-add-title.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+'use strict'
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		// Adding a new column to the 'resources' table
+		await queryInterface.addColumn('resources', 'title', {
+			type: Sequelize.STRING,
+			allowNull: true,
+		})
```
Comment (priyanka-TL) [inline src/database/migrations/20240531070905-alter-resource-add-title.js:10]: @adithyadinesh0412 add index to title
await queryInterface.addIndex('resources', ['title']);

File: src/database/models/resources.js [hunk 2]
```diff
@@ -12,6 +12,10 @@ module.exports = (sequelize, DataTypes) => {
 				allowNull: false,
 				type: DataTypes.STRING,
 			},
+			title: {
+				allowNull: true,
+				type: DataTypes.STRING,
+			},
```
Comment (priyanka-TL) [inline src/database/models/resources.js:18]: @adithyadinesh0412 add index to this field

File: src/services/resource.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	 */
+
+	static async list(user_id, queryParams, pageNo, limit) {
+		try {
+			let result = {
+				data: [],
+				count: 0,
+			}
+			let sort = {
+				sort_by: 'created_at',
```
Comment (priyanka-TL) [inline src/services/resource.js]: @adithyadinesh0412 add created_at, filter, type, status, sort_by etc in common

File: src/services/resource.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+				}
+			} else {
+				if ('type' in queryParams && queryParams.type.length > 0) {
+					filter.user_id = user_id
+					filter.type = queryParams.type
+				}
+
+				if ('status' in queryParams && queryParams.status.length > 0) {
+					filter.user_id = user_id
+					filter.status = common.STATUS_ENUM[common.STATUS_ENUM.indexOf(queryParams.status.toUpperCase())]
```
Comment (priyanka-TL) [inline src/services/resource.js]: @adithyadinesh0412 i think we dont need to check the status is exist in our db. If the status is wrong they will get empty data, this validation we can remove

File: src/services/resource.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+				sort_by: 'created_at',
+				order: 'DESC',
+			}
+			let filter = {}
+			if ('filter' in queryParams && queryParams.filter.toLowerCase() === common.FILTER_ALL.toLowerCase()) {
+				filter = {
+					user_id,
+				}
+			} else {
+				if ('type' in queryParams && queryParams.type.length > 0) {
```
Comment (priyanka-TL) [inline src/services/resource.js]: @adithyadinesh0412 at a time they can select only one type or all types, hence the length check also we can remove

File: src/services/resource.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				'sort_by' in queryParams &&
+				'sort_order' in queryParams &&
+				queryParams.sort_by.length > 0 &&
+				queryParams.sort_order.length > 0
+			) {
+				sort.sort_by = queryParams.sort_by
+				sort.order = queryParams.sort_order.toLowerCase() == 'desc' ? 'DESC' : 'ASC'
+			}
+
+			if (Object.keys(filter).length === 0) {
```
Comment (priyanka-TL) [inline src/services/resource.js]: @adithyadinesh0412 all the filters are not mandatory so we dont need to send this error message,

File: src/services/resource.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+					statusCode: httpStatusCode.ok,
+					message: 'RESOURCE_LISTED_SUCCESSFULLY',
+					result,
+				})
+			}
+
+			const uniqueOrganizationIds = [...new Set(resources.map((item) => item.organization_id))]
+			const orgDetailsResponse = await userRequests.listOrganization(uniqueOrganizationIds)
+			let orgDetails = {}
+			if (orgDetailsResponse.success) {
```
Comment (priyanka-TL) [inline src/services/resource.js]: @adithyadinesh0412 check the result is empty or not

File: src/services/resource.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+					return acc
+				}, {})
+			}
+
+			result.data = resources.map((res) => {
+				res.organization = orgDetails[res.organization_id] ? orgDetails[res.organization_id] : {}
+				delete res.organization_id
+				return res
+			})
+			result.count = result.data.length
```
Comment (priyanka-TL) [inline src/services/resource.js:93]: @adithyadinesh0412 is the pagination is working fine ?

**Review by priyanka-TL — CHANGES_REQUESTED** [review]

File: src/services/resource.js (hunk 7 above)
Comment (adithyadinesh0412) [reply src/services/resource.js]: Should we throw error if org details are not present? @priyanka-TL

File: src/constants/common.js [hunk 9]
```diff
@@ -55,6 +55,17 @@ module.exports = {
 	STATUS_ACTIVE: 'ACTIVE',
 	CREATED_BY_SYSTEM: 0,
 	PROJECT: 'project',
+	FILTER_ALL: 'ALL',
+	SORT_DESC: 'DESC',
+	SORT_ASC: 'ASC',
+	CREATED_AT: 'created_at',
+	QUERY_PARAMS: {
```
Comment (ankit1shahu) [inline src/constants/common.js]: Remove Query_Params @adithyadinesh0412

**Review by ankit1shahu — CHANGES_REQUESTED** [review]
Reviewed On 6June 12 AM

**Review by ankit1shahu — APPROVED** [review]

**Pull request merged**